File: JavaScript/6js 2/numberOfPairs.js
/*
|--------------------------------------------------------------------------
| Pair of gloves
|--------------------------------------------------------------------------
|
*/

function numberOfPairs(gloves) {
   const colorCount = gloves.reduce((count, color) => {
      count[color] = (count[color] || 0) + 1;
      return count;
   }, {});
   return Object.values(colorCount).reduce((pairs, count) => pairs + Math.floor(count / 2), 0);
}


function numberOfPairs(gloves) {
   return [...new Set(gloves)].reduce(
      (acc, el) => acc + ~~(gloves.filter((x) => x === el).length / 2),
      0
   );
}


function numberOfPairs(gloves) {
   let counter = 0;
   let copyGloves = [...gloves].sort();
   for (let i = 0; i < copyGloves.length; i++) {
      if (copyGloves[i] === copyGloves[i + 1]) {
         counter++;
         copyGloves.splice(i, 2);
         i--;
      }
   }
   return counter;
}


console.log(numberOfPairs(["red", "green", "red", "blue", "blue"]), 2);
console.log(numberOfPairs(["red", "red", "red", "red", "red", "red"]), 3);

File: Python/6py 2/number_of_pairs.py
"""
|--------------------------------------------------------------------------
| Pair of gloves
|--------------------------------------------------------------------------
|
"""

from collections import Counter
def number_of_pairs(gloves):
   color_count = Counter(gloves)
   return sum(count // 2 for count in color_count.values())


def number_of_pairs(gloves):
   return sum(gloves.count(color)//2 for color in set(gloves))


def number_of_pairs(gloves):
   result = 0
   for glove in set(gloves):
      result += gloves.count(glove) // 2
   return result


print(number_of_pairs(["red","red"]),1)
print(number_of_pairs(["red","green","blue"]),0)
print(number_of_pairs(["gray","black","purple","purple","gray","black"]),3)
print(number_of_pairs([]),0)
print(number_of_pairs(["red","green","blue","blue","red","green","red","red","red"]),4)